src/engine/Union.cpp

```diff
@@ -233,4 +237,73 @@ void Union::computeUnion(
     writeColumn(left, targetColumn, leftCol, 0u);
     writeColumn(right, targetColumn, rightCol, left.size());
   }
+  return res;
+}
+
+// _____________________________________________________________________________
+template <bool left>
+std::vector<size_t> Union::computePermutation() const {
+  size_t startOfUndefColumns = _subtrees[left ? 0 : 1]->getResultWidth();
+  std::vector<size_t> permutation(_columnOrigins.size());
+  for (size_t targetColIdx = 0; targetColIdx < _columnOrigins.size();
+       ++targetColIdx) {
+    size_t originIndex = _columnOrigins.at(targetColIdx)[left ? 0 : 1];
+    if (originIndex == NO_COLUMN) {
+      originIndex = startOfUndefColumns;
+      startOfUndefColumns++;
+    }
+    permutation[targetColIdx] = originIndex;
+  }
+  return permutation;
+}
+
+// _____________________________________________________________________________
+IdTable Union::transformToCorrectColumnFormat(
+    IdTable idTable, const std::vector<size_t>& permutation) const {
+  while (idTable.numColumns() < getResultWidth()) {
+    idTable.addEmptyColumn();
+    std::ranges::fill(idTable.getColumn(idTable.numColumns() - 1),
+                      Id::makeUndefined());
+  }
+
+  // i + 1 because once everything expect the last column is in the correct
+  // position, the last column is already in the correct position so the last
+  // iteration would always swap the last column with itself.
+  for (size_t i = 0; i + 1 < permutation.size(); ++i) {
+    size_t ind = permutation[i];
+    while (ind < i) {
+      ind = permutation[ind];
+    }
+    idTable.swapColumns(i, ind);
+  }
+  return idTable;
+}
```

Comment:
I am pretty sure that the IdTable class has a `setColumnSubset` function or something, that can be used to apply a permutation, saving you the logic here + testing it thoroughly.

Reply:
Seems like it reallocates unnecessary memory (by creating a second vector/array and replacing the original one) instead of swapping the elements in place like it's done here, but it definitely works.